Library/Homebrew/cmd/outdated.rb

```diff
@@ -20,46 +22,79 @@ def outdated_args
-                          "value for <version> is `v1`. See the docs for examples of using the JSON "\
-                          "output: <https://docs.brew.sh/Querying-Brew>"
+             description: "Print output in JSON format. There are two versions: v1 and v2. " \
+                          "v1 is deprecated and is currently the default if no version is specified. " \
+                          "v2 prints outdated formulae and casks. "
       switch "--fetch-HEAD",
              description: "Fetch the upstream repository to detect if the HEAD installation of the "\
                           "formula is outdated. Otherwise, the repository's HEAD will only be checked for "\
                           "updates when a new stable or development version has been released."
+      switch "--greedy",
+             description: "Print outdated casks with `auto_updates` or `version :latest`."
+      switch "--formula",
+             description: "Treat all arguments as formulae."
+      switch "--cask",
+             description: "Treat all arguments as casks."
       switch :debug
       conflicts "--quiet", "--verbose", "--json"
+      conflicts "--formula", "--cask"
     end
   end
 
   def outdated
     outdated_args.parse
 
-    formulae = if args.resolved_formulae.blank?
-      Formula.installed
-    else
-      args.resolved_formulae
-    end
-    if args.json
-      raise UsageError, "invalid JSON version: #{args.json}" unless ["v1", true].include? args.json
+    case json_version
+    when :v1
+      opoo "JSON v1 has been deprecated; please use --json=v2"
+
+      outdated = if args.formula? || !args.cask?
+        outdated_formulae
+      else
+        outdated_casks
+      end
+
+      puts JSON.generate(json_info(outdated))
+
+    when :v2
+      formulae, casks = if args.formula?
+        [outdated_formulae, []]
+      elsif args.cask?
+        [[], outdated_casks]
+      else
+        outdated_formulae_casks
+      end
+
+      puts JSON.generate({
+                           "formulae" => json_info(formulae),
+                           "casks"    => json_info(casks),
+                         })
+
+      outdated = formulae + casks
 
-      outdated = print_outdated_json(formulae)
     else
-      outdated = print_outdated(formulae)
+      outdated = if args.formula?
+        outdated_formulae
+      elsif args.cask?
+        outdated_casks
+      else
+        outdated_formulae_casks.flatten
+      end
+
+      print_outdated(outdated)
     end
-    Homebrew.failed = args.resolved_formulae.present? && !outdated.empty?
+
+    Homebrew.failed = args.named.present? && outdated.present?
   end
 
-  def print_outdated(formulae)
-    verbose = ($stdout.tty? || args.verbose?) && !args.quiet?
-    fetch_head = args.fetch_HEAD?
+  def print_outdated(formula_or_cask)
+    return formula_or_cask.each { |f_or_c| print_outdated(f_or_c) } if formula_or_cask.is_a? Array
```

Comment:
Might be nicer to do `Array(formula_or_cask)` and unconditionally loop (even if there's only one value)